libs/zzz/solver/src/calc.ts

```diff
@@ -13,22 +17,58 @@ import {
 import type { Stats } from '@genshin-optimizer/zzz/db'
 import type { DiscStats } from './common'
 
+export function passSetFilter(
+  discs: DiscStats[],
+  filter2: DiscSetKey[],
+  filter4: DiscSetKey[]
+): boolean {
+  const setCount: Partial<Record<DiscSetKey, number>> = {}
+  if (!filter4.length && !filter2.length) return true
+  for (const { setKey } of discs) setCount[setKey] = (setCount[setKey] || 0) + 1
+
+  if (filter4.length) {
+    let has4 = false
+    for (const key in setCount) {
+      const k = key as DiscSetKey
+      if (setCount[k]! >= 4 && filter4.includes(k)) {
+        has4 = true
```

Comment:
nit: we seem to do more of arr.every() or arr.any() checks rather than this for-break paradigm. not a big deal tho, just weird to read

Reply:
I use more for loops in calc because they are more performant (how much so I can't attest), I much rather use array function otherwise for robustness